### Objective: Loading → inspecting → cleaning → validating → merging → saving the master dataset

In [2]:
import pandas as pd
import os
data_path = "../data"
files = os.listdir(data_path)
for file in files:
    file_path = os.path.join(data_path, file)
    print("\n" + "=" * 80)
    print(file)
    print("=" * 80)
    # Read without assuming a header
    raw = pd.read_excel(file_path, header=None)
    # Show first 10 rows
    display(raw.head(10))


DS11_Nitrogen_Thesis.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,DS11 — Leaf Nitrogen (N) Content | Kjeldahl Me...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Category: F — Leaf tissue / nutrient analysis ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Plot ID,Village,Season,Sample date,Growth stage,Leaf position,N content pct DW,N class,Kjeldahl N mg kg,Total protein pct,Chlorophyll SPAD,N fertiliser kg ha,Method,Lab ID
4,WC-001,Matiaria,2021-22,2022-06-01,Grand_growth,4th_leaf,2.896,Excess,28960,18.1,54.1,150,Kjeldahl_digestion,ICAR-LEAF-N-0001
5,WC-002,Sirsa,2023-24,2022-06-04,Grand_growth,4th_leaf,1.499,Deficient,14990,9.369,28.8,180,Kjeldahl_digestion,ICAR-LEAF-N-0002
6,WC-003,Matiaria,2023-24,2022-06-07,Tillering,4th_leaf,1.616,Adequate,16160,10.1,31,200,Kjeldahl_digestion,ICAR-LEAF-N-0003
7,WC-004,Matiaria,2022-23,2022-06-10,Tillering,TVD,1.509,Adequate,15090,9.431,30.1,200,Kjeldahl_digestion,ICAR-LEAF-N-0004
8,WC-005,Sirsa,2023-24,2022-06-13,Grand_growth,TVD,3.01,Excess,30100,18.812,48.5,200,Kjeldahl_digestion,ICAR-LEAF-N-0005
9,WC-006,Sirsa,2021-22,2022-06-16,Tillering,TVD,1.402,Deficient,14020,8.762,26.8,120,Kjeldahl_digestion,ICAR-LEAF-N-0006



DS12_Phosphorus_Thesis.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,DS12 — Leaf Phosphorus (P) Content | Vanadomol...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Category: F — Leaf tissue / nutrient analysis ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Plot ID,Village,Season,Sample date,Growth stage,P content pct DW,P class,P mg per kg DW,P2O5 pct,Available P Olsen,P fertiliser kg ha,P uptake kg ha,Method,Lab ID
4,WC-001,Matiaria,2021-22,2022-06-01,Grand_growth,0.2241,Adequate,2241,0.5134,19.56,50,13.44,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0001
5,WC-002,Sirsa,2023-24,2022-06-04,Grand_growth,0.2157,Adequate,2157,0.4942,30.41,40,9.77,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0002
6,WC-003,Matiaria,2023-24,2022-06-07,Tillering,0.3566,Excess,3566,0.817,51.26,40,25.94,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0003
7,WC-004,Matiaria,2022-23,2022-06-10,Tillering,0.3,Excess,3000,0.6873,42.03,80,17.37,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0004
8,WC-005,Sirsa,2023-24,2022-06-13,Grand_growth,0.2103,Adequate,2103,0.4818,39.11,80,14.27,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0005
9,WC-006,Sirsa,2021-22,2022-06-16,Tillering,0.274,Adequate,2740,0.6277,28.74,50,22.82,Vanadomolybdate_colorimetry,ICAR-LEAF-P-0006



DS13_Potassium_Thesis.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,DS13 — Leaf Potassium (K) Content | Flame Phot...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Category: F — Leaf tissue / nutrient analysis ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Plot ID,Village,Season,Sample date,Growth stage,K content pct DW,K class,K mg per kg DW,K2O pct,Exchangeable K meq,K fertiliser kg ha,K Na ratio,Method,Lab ID
4,WC-001,Matiaria,2021-22,2022-06-01,Grand_growth,1.482,Adequate,14820,1.786,0.941,60,11.98,Flame_photometry,ICAR-LEAF-K-0001
5,WC-002,Sirsa,2023-24,2022-06-04,Grand_growth,1.29,Adequate,12900,1.554,0.776,80,8.59,Flame_photometry,ICAR-LEAF-K-0002
6,WC-003,Matiaria,2023-24,2022-06-07,Tillering,1.843,Adequate,18430,2.221,0.207,80,14.4,Flame_photometry,ICAR-LEAF-K-0003
7,WC-004,Matiaria,2022-23,2022-06-10,Tillering,1.906,Adequate,19060,2.297,0.848,100,10.71,Flame_photometry,ICAR-LEAF-K-0004
8,WC-005,Sirsa,2023-24,2022-06-13,Grand_growth,2.044,Adequate,20440,2.463,0.657,60,13.03,Flame_photometry,ICAR-LEAF-K-0005
9,WC-006,Sirsa,2021-22,2022-06-16,Tillering,1.612,Adequate,16120,1.942,0.862,60,6.54,Flame_photometry,ICAR-LEAF-K-0006



DS14_GNDVI_Derived_Thesis.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,DS14 — GNDVI & Vegetation Indices (Derived) | ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Category: G — Derived / computed variables | ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Plot ID,Village,Season,Acquisition date,Growth stage,GNDVI,GNDVI class,NDVI,EVI,LSWI,Canopy LAI est,Chlorophyll est ug cm2,Formula,Source dataset
4,WC-001,Matiaria,2021-22,2022-01-10,Tillering,0.6253,High_vigour,0.8474,1.4089,0.1474,3.34,31.53,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands
5,WC-002,Sirsa,2023-24,2022-01-12,Grand_growth,0.8444,High_vigour,0.7426,0.6684,0.6517,5.06,38.99,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands
6,WC-003,Matiaria,2023-24,2022-01-14,Germination,0.3347,Low_vigour,0.7982,4.982,0.0379,2.04,12.91,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands
7,WC-004,Matiaria,2022-23,2022-01-16,Tillering,0.783,High_vigour,0.4611,0.3458,0.2131,4.58,35.75,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands
8,WC-005,Sirsa,2023-24,2022-01-18,Ripening,0.6665,High_vigour,0.8019,1.3546,0.2645,3.6,27.91,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands
9,WC-006,Sirsa,2021-22,2022-01-20,Tillering,0.5805,High_vigour,0.6585,0.9979,0.4106,3.1,27,GNDVI=(NIR-Green)/(NIR+Green),DS4_Landsat_Bands



DS17_Chlorophyll_Thesis.xlsx


,0,1,2,3,4,5,6,7,8,9,...,24,25,26,27,28,29,30,31,32,33
0,DS17 — Chlorophyll Content & Fluorescence | We...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Category: F — Leaf tissue / nutrient analysis ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Parameters: Chl_a · Chl_b · Total_Chl · Chl_a:...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Source: ICAR crop lab — Acetone/DMSO extractio...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Role in model: Canopy health indicator — links...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,Plot ID,Village,Cultivar,Cycle,Season,Block,Sample date,Time of sampling,Growth stage,Leaf position,...,Chl specific absorption,Light use efficiency LUE,GNDVI from DS14,N content pct from DS11,Nutrient balance NB DS16,Stress type,Chl status class,Extraction method,Lab ID,Notes
7,WC-001,Matiaria,CoVSI9,Plant,2021-22,2,2022-06-01,17:00,Grand_growth,4th_leaf,...,0.12487,0.01773,0.62527,2.896,0.53153,Pest,High,Ethanol_96pct,ICAR-CHL-DS17-0001,NaN
8,WC-002,Sirsa,Co0238,Ratoon1,2023-24,4,2022-06-04,11:00,Grand_growth,4th_leaf,...,0.13028,0.01752,0.84435,1.499,0.39002,Nutrient_deficiency,High,Acetone_80pct,ICAR-CHL-DS17-0002,NaN
9,WC-003,Matiaria,CoS767,Ratoon2,2023-24,3,2022-06-07,15:00,Tillering,4th_leaf,...,0.1018,0.01477,0.33467,1.616,0.54277,Drought,Adequate,Ethanol_96pct,ICAR-CHL-DS17-0003,NaN



DS1_plot_registry.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12
0,DS1 — Experimental Plot Registry | West Champa...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Plot_ID,Village,Block,Season,Latitude_deg,Longitude_deg,Plot_Area_ha,Irrigation_type,Soil_type,Planting_date,Harvest_date,Growth_days,Notes
2,WC-001,Matiaria,2,2021-22,26.87983,84.67516,0.4,Canal,Sandy_loam,2021-11-01,2022-10-15,344,NaN
3,WC-002,Sirsa,4,2023-24,27.03752,84.74989,1.435,Tubewell,Clay_loam,2021-11-04,2022-10-18,388,NaN
4,WC-003,Matiaria,3,2023-24,27.03765,84.74865,1.55,Canal,Sandy_loam,2021-11-07,2022-10-21,355,NaN
5,WC-004,Matiaria,1,2022-23,26.94336,84.57217,1.915,Canal,Alluvial,2021-11-10,2022-10-24,355,NaN
6,WC-005,Sirsa,2,2023-24,26.95796,84.65759,1.193,Tubewell,Sandy_loam,2021-11-13,2022-10-27,376,NaN
7,WC-006,Sirsa,3,2021-22,26.91211,84.72791,0.493,Canal,Alluvial,2021-11-16,2022-10-30,361,NaN
8,WC-007,Matiaria,2,2022-23,26.8733,84.68986,0.853,Canal,Alluvial,2021-11-19,2022-11-02,339,NaN
9,WC-008,Masjidwa,3,2021-22,26.88217,84.44894,1.035,Tubewell,Clay_loam,2021-11-22,2022-11-05,345,NaN



DS2_TSH_Records.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,DS2 — TSH & Sugar Quality Records | Response V...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Plot_ID,Village,Season,Cultivar,Cycle,Block,TCH_t_per_ha,CCS_pct,TSH_t_per_ha,Sugar_recovery_pct,Juice_purity_pct,Brix_pct,Pol_pct,Mill_ID
2,WC-001,Matiaria,2021-22,CoVSI9,Plant,2,11.18,11.8,1.023,7.71,84.18,19.36,20.64,WC_Mill_03
3,WC-002,Sirsa,2023-24,Co0238,Ratoon1,4,104.19,11.42,9.136,7.76,86.65,20.42,20.91,WC_Mill_03
4,WC-003,Matiaria,2023-24,CoS767,Ratoon2,3,71.81,10.75,6.72,7.83,86.83,21.26,16.78,WC_Mill_02
5,WC-004,Matiaria,2022-23,Co0238,Ratoon2,1,51.5,11.93,4.583,8.24,92.49,22.75,15.22,WC_Mill_01
6,WC-005,Sirsa,2023-24,Co0238,Ratoon1,2,88.41,10.06,7.989,9.6,88.12,21.29,20.18,WC_Mill_01
7,WC-006,Sirsa,2021-22,Co86032,Ratoon1,3,136.25,13.09,15.845,9.39,90.68,21.68,15.74,WC_Mill_01
8,WC-007,Matiaria,2022-23,Co0238,Ratoon2,2,17.69,12.82,1.824,8.93,83.25,19.36,17.2,WC_Mill_01
9,WC-008,Masjidwa,2021-22,CoVSI9,Ratoon2,3,29.13,10.77,3.383,8.12,84.71,23.08,15.03,WC_Mill_01



DS6_Precipitaion.xlsx


,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
0,DS6 — Accumulated Precipitation Data | IMD + E...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Plot_ID,Village,Season,Source,Grid_lat,Grid_lon,Precip_growth_phase_mm,Precip_Jun_Aug_mm,Precip_Sep_Nov_mm,Precip_Dec_Feb_mm,Dry_spell_days_gt5,Wet_days_gt10mm,Max_1day_precip_mm,Drought_flag,Excess_rain_flag,SPI_3month
2,WC-001,Matiaria,2021-22,ERA5_0.25deg,27.01,84.64,1066.3,650,286.3,47.4,10,69,45.2,No,No,-0.465
3,WC-002,Sirsa,2023-24,IMD_gridded,26.81,84.46,958.6,609.9,271.3,57.3,4,76,168.5,No,No,0.612
4,WC-003,Matiaria,2023-24,ERA5_0.25deg,27.01,84.61,923.4,548,258.8,51.1,4,44,134.2,No,No,1.638
5,WC-004,Matiaria,2022-23,IMD_gridded,26.92,84.49,710.1,424.2,225.7,35.7,14,40,178.7,No,No,1.043
6,WC-005,Sirsa,2023-24,IMD_gridded,26.85,84.75,955.9,573,265.5,46.4,28,51,163.6,No,No,1.362
7,WC-006,Sirsa,2021-22,IMD_gridded,26.83,84.46,913.3,592.2,283.2,46.7,17,83,106.2,No,No,0.31
8,WC-007,Matiaria,2022-23,ERA5_0.25deg,27.04,84.46,896.8,555.7,244.8,47.4,10,53,112.8,No,No,0.522
9,WC-008,Masjidwa,2021-22,IMD_gridded,26.82,84.72,886.3,535.7,266.7,34.1,25,30,66.7,No,No,0.256


In [3]:
# Correct header rows identified from the raw Excel sheets

header_rows = {
    "DS1_plot_registry.xlsx": 1,
    "DS2_TSH_Records.xlsx": 1,
    "DS6_Precipitaion.xlsx": 1,
    "DS11_Nitrogen_Thesis.xlsx": 3,
    "DS12_Phosphorus_Thesis.xlsx": 3,
    "DS13_Potassium_Thesis.xlsx": 3,
    "DS14_GNDVI_Derived_Thesis.xlsx": 3,
    "DS17_Chlorophyll_Thesis.xlsx": 6
}

datasets = {}

for file, header_row in header_rows.items():
    file_path = os.path.join(data_path, file)
    datasets[file] = pd.read_excel(file_path, header=header_row)

print("All datasets loaded with correct headers.")

All datasets loaded with correct headers.


In [4]:
for name, df in datasets.items():
    print("\n" + "=" * 70)
    print(name)
    print("=" * 70)
    print(df.columns.tolist())


DS1_plot_registry.xlsx
['Plot_ID', 'Village', 'Block', 'Season', 'Latitude_deg', 'Longitude_deg', 'Plot_Area_ha', 'Irrigation_type', 'Soil_type', 'Planting_date', 'Harvest_date', 'Growth_days', 'Notes']

DS2_TSH_Records.xlsx
['Plot_ID', 'Village', 'Season', 'Cultivar', 'Cycle', 'Block', 'TCH_t_per_ha', 'CCS_pct', 'TSH_t_per_ha', 'Sugar_recovery_pct', 'Juice_purity_pct', 'Brix_pct', 'Pol_pct', 'Mill_ID']

DS6_Precipitaion.xlsx
['Plot_ID', 'Village', 'Season', 'Source', 'Grid_lat', 'Grid_lon', 'Precip_growth_phase_mm', 'Precip_Jun_Aug_mm', 'Precip_Sep_Nov_mm', 'Precip_Dec_Feb_mm', 'Dry_spell_days_gt5', 'Wet_days_gt10mm', 'Max_1day_precip_mm', 'Drought_flag', 'Excess_rain_flag', 'SPI_3month']

DS11_Nitrogen_Thesis.xlsx
['Plot ID', 'Village', 'Season', 'Sample date', 'Growth stage', 'Leaf position', 'N content pct DW', 'N class', 'Kjeldahl N mg kg', 'Total protein pct', 'Chlorophyll SPAD', 'N fertiliser kg ha', 'Method', 'Lab ID']

DS12_Phosphorus_Thesis.xlsx
['Plot ID', 'Village', 'Seaso

In [5]:
for name, df in datasets.items():

    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)

    print("Shape:", df.shape)

    # Missing values
    missing = df.isna().sum()
    missing = missing[missing > 0]

    print("\nMissing values:")
    if len(missing) == 0:
        print("No missing values")
    else:
        print(missing)

    # Duplicate rows
    print("\nDuplicate rows:", df.duplicated().sum())

    # Plot ID information
    plot_col = "Plot_ID" if "Plot_ID" in df.columns else "Plot ID"

    print("Unique Plot IDs:", df[plot_col].nunique())
    print("Total rows:", len(df))


DS1_plot_registry.xlsx
Shape: (120, 13)

Missing values:
Notes    120
dtype: int64

Duplicate rows: 0
Unique Plot IDs: 120
Total rows: 120

DS2_TSH_Records.xlsx
Shape: (120, 14)

Missing values:
No missing values

Duplicate rows: 0
Unique Plot IDs: 120
Total rows: 120

DS6_Precipitaion.xlsx
Shape: (120, 16)

Missing values:
No missing values

Duplicate rows: 0
Unique Plot IDs: 120
Total rows: 120

DS11_Nitrogen_Thesis.xlsx
Shape: (121, 14)

Missing values:
Village               1
Season                1
Sample date           1
Growth stage          1
Leaf position         1
N content pct DW      1
N class               1
Kjeldahl N mg kg      1
Total protein pct     1
Chlorophyll SPAD      1
N fertiliser kg ha    1
Method                1
Lab ID                1
dtype: int64

Duplicate rows: 0
Unique Plot IDs: 121
Total rows: 121

DS12_Phosphorus_Thesis.xlsx
Shape: (121, 14)

Missing values:
Village               1
Season                1
Sample date           1
Growth stage          

In [6]:
plot_id_sets = {}

for name, df in datasets.items():

    plot_col = "Plot_ID" if "Plot_ID" in df.columns else "Plot ID"

    plot_id_sets[name] = set(df[plot_col].dropna().astype(str).str.strip())

print("Plot ID consistency check\n")

for name, ids in plot_id_sets.items():
    print(f"{name}: {len(ids)} unique Plot IDs")

print("\n" + "=" * 80)
print("COMMON PLOT IDs ACROSS ALL DATASETS")
print("=" * 80)

common_ids = set.intersection(*plot_id_sets.values())

print("Number of common Plot IDs:", len(common_ids))
print("First 20 common IDs:", sorted(common_ids)[:20])

Plot ID consistency check

DS1_plot_registry.xlsx: 120 unique Plot IDs
DS2_TSH_Records.xlsx: 120 unique Plot IDs
DS6_Precipitaion.xlsx: 120 unique Plot IDs
DS11_Nitrogen_Thesis.xlsx: 121 unique Plot IDs
DS12_Phosphorus_Thesis.xlsx: 121 unique Plot IDs
DS13_Potassium_Thesis.xlsx: 121 unique Plot IDs
DS14_GNDVI_Derived_Thesis.xlsx: 121 unique Plot IDs
DS17_Chlorophyll_Thesis.xlsx: 101 unique Plot IDs

COMMON PLOT IDs ACROSS ALL DATASETS
Number of common Plot IDs: 100
First 20 common IDs: ['WC-001', 'WC-002', 'WC-003', 'WC-004', 'WC-005', 'WC-006', 'WC-007', 'WC-008', 'WC-009', 'WC-010', 'WC-011', 'WC-012', 'WC-013', 'WC-014', 'WC-015', 'WC-016', 'WC-017', 'WC-018', 'WC-019', 'WC-020']


In [7]:
# Compare Plot IDs between every dataset and the common set

for name, ids in plot_id_sets.items():
    missing_from_common = ids - common_ids

    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)

    print("Plot IDs not present in ALL datasets:", len(missing_from_common))
    print(sorted(missing_from_common))


DS1_plot_registry.xlsx
Plot IDs not present in ALL datasets: 20
['WC-101', 'WC-102', 'WC-103', 'WC-104', 'WC-105', 'WC-106', 'WC-107', 'WC-108', 'WC-109', 'WC-110', 'WC-111', 'WC-112', 'WC-113', 'WC-114', 'WC-115', 'WC-116', 'WC-117', 'WC-118', 'WC-119', 'WC-120']

DS2_TSH_Records.xlsx
Plot IDs not present in ALL datasets: 20
['WC-101', 'WC-102', 'WC-103', 'WC-104', 'WC-105', 'WC-106', 'WC-107', 'WC-108', 'WC-109', 'WC-110', 'WC-111', 'WC-112', 'WC-113', 'WC-114', 'WC-115', 'WC-116', 'WC-117', 'WC-118', 'WC-119', 'WC-120']

DS6_Precipitaion.xlsx
Plot IDs not present in ALL datasets: 20
['WC-101', 'WC-102', 'WC-103', 'WC-104', 'WC-105', 'WC-106', 'WC-107', 'WC-108', 'WC-109', 'WC-110', 'WC-111', 'WC-112', 'WC-113', 'WC-114', 'WC-115', 'WC-116', 'WC-117', 'WC-118', 'WC-119', 'WC-120']

DS11_Nitrogen_Thesis.xlsx
Plot IDs not present in ALL datasets: 21
['Source: Leaf Nitrogen (N) Content | Kjeldahl Method | ICAR Crop Lab  |  Plot IDs: WC-001 to WC-120  |  Generated for thesis: Semiparame

In [8]:
all_plot_ids = sorted(set.union(*plot_id_sets.values()))

coverage = pd.DataFrame({
    "Plot_ID": all_plot_ids
})

for name, ids in plot_id_sets.items():
    short_name = name.replace(".xlsx", "")
    coverage[short_name] = coverage["Plot_ID"].isin(ids)

coverage["Dataset_Count"] = coverage.iloc[:, 1:].sum(axis=1)

display(coverage)

,Plot_ID,DS1_plot_registry,DS2_TSH_Records,DS6_Precipitaion,DS11_Nitrogen_Thesis,DS12_Phosphorus_Thesis,DS13_Potassium_Thesis,DS14_GNDVI_Derived_Thesis,DS17_Chlorophyll_Thesis,Dataset_Count
0,Note: 100 plot-level observations | Plot IDs W...,False,False,False,False,False,False,False,True,1
1,Source: GNDVI & Vegetation Indices (Derived) |...,False,False,False,False,False,False,True,False,1
2,Source: Leaf Nitrogen (N) Content | Kjeldahl M...,False,False,False,True,False,False,False,False,1
3,Source: Leaf Phosphorus (P) Content | Vanadomo...,False,False,False,False,True,False,False,False,1
4,Source: Leaf Potassium (K) Content | Flame Pho...,False,False,False,False,False,True,False,False,1
...,...,...,...,...,...,...,...,...,...,...
120,WC-116,True,True,True,True,True,True,True,False,7
121,WC-117,True,True,True,True,True,True,True,False,7
122,WC-118,True,True,True,True,True,True,True,False,7
123,WC-119,True,True,True,True,True,True,True,False,7


In [9]:
# Standardize column names and remove metadata rows

for name, df in datasets.items():

    # Remove leading/trailing spaces from column names
    df.columns = df.columns.str.strip()

    # Standardize Plot ID column name
    if "Plot ID" in df.columns:
        df.rename(columns={"Plot ID": "Plot_ID"}, inplace=True)

    # Keep only actual plot records
    datasets[name] = df[
        datasets[name]["Plot_ID"].astype(str).str.match(r"^WC-\d{3}$", na=False)
    ].copy()

print("Metadata rows removed and Plot_ID standardized.")

Metadata rows removed and Plot_ID standardized.


In [10]:
for name, df in datasets.items():
    print(f"{name}: {df.shape[0]} rows | {df['Plot_ID'].nunique()} unique Plot IDs")

DS1_plot_registry.xlsx: 120 rows | 120 unique Plot IDs
DS2_TSH_Records.xlsx: 120 rows | 120 unique Plot IDs
DS6_Precipitaion.xlsx: 120 rows | 120 unique Plot IDs
DS11_Nitrogen_Thesis.xlsx: 120 rows | 120 unique Plot IDs
DS12_Phosphorus_Thesis.xlsx: 120 rows | 120 unique Plot IDs
DS13_Potassium_Thesis.xlsx: 120 rows | 120 unique Plot IDs
DS14_GNDVI_Derived_Thesis.xlsx: 120 rows | 120 unique Plot IDs
DS17_Chlorophyll_Thesis.xlsx: 100 rows | 100 unique Plot IDs


In [11]:
plot_id_sets = {
    name: set(df["Plot_ID"])
    for name, df in datasets.items()
}

common_ids = set.intersection(*plot_id_sets.values())

print("Common Plot IDs across all 8 datasets:", len(common_ids))
print("Range:", min(common_ids), "to", max(common_ids))
print("First 10:", sorted(common_ids)[:10])
print("Last 10:", sorted(common_ids)[-10:])

Common Plot IDs across all 8 datasets: 100
Range: WC-001 to WC-100
First 10: ['WC-001', 'WC-002', 'WC-003', 'WC-004', 'WC-005', 'WC-006', 'WC-007', 'WC-008', 'WC-009', 'WC-010']
Last 10: ['WC-091', 'WC-092', 'WC-093', 'WC-094', 'WC-095', 'WC-096', 'WC-097', 'WC-098', 'WC-099', 'WC-100']


### Inspect missing values and data types

In [14]:
for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)
    print("Shape:", df.shape)
    # Missing values
    missing = df.isna().sum()
    missing = missing[missing > 0]
    print("\nMissing values:")
    if missing.empty:
        print("None")
    else:
        print(missing)
    # Data types
    print("\nData types:")
    print(df.dtypes)


DS1_plot_registry.xlsx
Shape: (120, 13)

Missing values:
Notes    120
dtype: int64

Data types:
Plot_ID             object
Village             object
Block                int64
Season              object
Latitude_deg       float64
Longitude_deg      float64
Plot_Area_ha       float64
Irrigation_type     object
Soil_type           object
Planting_date       object
Harvest_date        object
Growth_days          int64
Notes              float64
dtype: object

DS2_TSH_Records.xlsx
Shape: (120, 14)

Missing values:
None

Data types:
Plot_ID                object
Village                object
Season                 object
Cultivar               object
Cycle                  object
Block                   int64
TCH_t_per_ha          float64
CCS_pct               float64
TSH_t_per_ha          float64
Sugar_recovery_pct    float64
Juice_purity_pct      float64
Brix_pct              float64
Pol_pct               float64
Mill_ID                object
dtype: object

DS6_Precipitaion.xlsx
Shape: 

### Check the important numerical variables

In [15]:
for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)
    numeric_cols = df.select_dtypes(include="number").columns
    print("Numerical columns:")
    print(list(numeric_cols))
    if len(numeric_cols) > 0:
        display(df[numeric_cols].describe().T)


DS1_plot_registry.xlsx
Numerical columns:
['Block', 'Latitude_deg', 'Longitude_deg', 'Plot_Area_ha', 'Growth_days', 'Notes']


,count,mean,std,min,25%,50%,75%,max
Block,120.0,2.516667,1.122597,1.00000,2.000000,2.00000,4.000000,4.00000
Latitude_deg,120.0,26.933748,0.078348,26.80284,26.871133,26.93903,27.003955,27.04824
Longitude_deg,120.0,84.536985,0.113069,84.35582,84.434370,84.53466,84.622770,84.74989
Plot_Area_ha,120.0,1.325558,0.646672,0.27700,0.764000,1.29600,1.916250,2.47700
Growth_days,120.0,361.791667,19.286820,332.00000,344.000000,361.00000,378.000000,394.00000
Notes,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN



DS2_TSH_Records.xlsx
Numerical columns:
['Block', 'TCH_t_per_ha', 'CCS_pct', 'TSH_t_per_ha', 'Sugar_recovery_pct', 'Juice_purity_pct', 'Brix_pct', 'Pol_pct']


,count,mean,std,min,25%,50%,75%,max
Block,120.0,2.516667,1.122597,1.000,2.0000,2.0000,4.0000,4.000
TCH_t_per_ha,120.0,51.061583,37.982855,2.290,22.0500,45.5700,70.4175,218.390
CCS_pct,120.0,11.400167,1.130745,9.500,10.4175,11.3800,12.3025,13.450
TSH_t_per_ha,120.0,5.227150,3.888945,0.217,2.1840,4.7655,7.0070,21.023
Sugar_recovery_pct,120.0,9.213333,1.012821,7.710,8.3275,9.0300,10.1225,11.100
Juice_purity_pct,120.0,87.954167,3.602240,82.080,84.7075,87.9400,90.8975,93.890
Brix_pct,120.0,21.005750,1.704704,18.110,19.5300,20.9450,22.5050,23.980
Pol_pct,120.0,18.313083,2.178829,15.000,16.5250,18.1350,20.3175,21.940



DS6_Precipitaion.xlsx
Numerical columns:
['Grid_lat', 'Grid_lon', 'Precip_growth_phase_mm', 'Precip_Jun_Aug_mm', 'Precip_Sep_Nov_mm', 'Precip_Dec_Feb_mm', 'Dry_spell_days_gt5', 'Wet_days_gt10mm', 'Max_1day_precip_mm', 'SPI_3month']


,count,mean,std,min,25%,50%,75%,max
Grid_lat,120.0,26.918417,0.070389,26.800,26.8600,26.920,26.9800,27.050
Grid_lon,120.0,84.554500,0.118484,84.360,84.4600,84.580,84.6500,84.750
Precip_growth_phase_mm,120.0,963.338333,221.880533,356.000,839.9000,957.350,1095.5250,1586.700
Precip_Jun_Aug_mm,120.0,597.052500,134.300305,197.500,522.9750,587.400,665.7500,987.900
Precip_Sep_Nov_mm,120.0,271.255833,65.678725,94.700,235.7750,266.250,311.4500,456.000
Precip_Dec_Feb_mm,120.0,48.105833,12.061846,17.600,40.5000,48.450,55.9500,78.300
Dry_spell_days_gt5,120.0,14.200000,8.391272,0.000,7.0000,14.000,21.0000,29.000
Wet_days_gt10mm,120.0,59.975000,16.441116,30.000,47.0000,61.000,73.0000,89.000
Max_1day_precip_mm,120.0,114.538333,40.130795,41.400,83.9000,117.250,147.1500,179.900
SPI_3month,120.0,0.070850,0.976531,-2.117,-0.5605,0.096,0.7425,2.532



DS11_Nitrogen_Thesis.xlsx
Numerical columns:
['N content pct DW', 'Kjeldahl N mg kg', 'Total protein pct', 'Chlorophyll SPAD', 'N fertiliser kg ha']


,count,mean,std,min,25%,50%,75%,max
N content pct DW,120.0,2.085092,0.448238,0.911,1.74450,2.1045,2.3725,3.171
Kjeldahl N mg kg,120.0,20850.916667,4482.381944,9110.000,17445.00000,21045.0000,23725.0000,31710.000
Total protein pct,120.0,13.031800,2.801527,5.694,10.90275,13.1530,14.8280,19.819
Chlorophyll SPAD,120.0,37.289167,8.469052,19.000,30.40000,37.7000,43.0500,58.100
N fertiliser kg ha,120.0,150.833333,37.787182,100.000,120.00000,150.0000,180.0000,200.000



DS12_Phosphorus_Thesis.xlsx
Numerical columns:
['P content pct DW', 'P mg per kg DW', 'P2O5 pct', 'Available P Olsen', 'P fertiliser kg ha', 'P uptake kg ha']


,count,mean,std,min,25%,50%,75%,max
P content pct DW,120.0,0.248498,0.065083,0.0616,0.208875,0.24270,0.290325,0.4170
P mg per kg DW,120.0,2484.983333,650.833699,616.0000,2088.750000,2427.00000,2903.250000,4170.0000
P2O5 pct,120.0,0.569307,0.149104,0.1411,0.478500,0.55605,0.665100,0.9553
Available P Olsen,120.0,32.159917,13.234880,8.0600,20.772500,32.35500,44.645000,53.5100
P fertiliser kg ha,120.0,56.000000,15.415805,40.0000,40.000000,50.00000,60.000000,80.0000
P uptake kg ha,120.0,16.415333,5.084457,3.7400,12.975000,15.92000,19.740000,29.7200



DS13_Potassium_Thesis.xlsx
Numerical columns:
['K content pct DW', 'K mg per kg DW', 'K2O pct', 'Exchangeable K meq', 'K fertiliser kg ha', 'K Na ratio']


,count,mean,std,min,25%,50%,75%,max
K content pct DW,120.0,1.727500,0.414866,0.852,1.37850,1.7605,1.95700,2.810
K mg per kg DW,120.0,17275.000000,4148.664461,8520.000,13785.00000,17605.0000,19570.00000,28100.000
K2O pct,120.0,2.081625,0.499916,1.027,1.66150,2.1215,2.35800,3.386
Exchangeable K meq,120.0,0.556117,0.215462,0.166,0.39075,0.5690,0.72725,0.943
K fertiliser kg ha,120.0,89.000000,22.729687,60.000,60.00000,80.0000,105.00000,120.000
K Na ratio,120.0,9.837833,3.794201,3.640,7.17000,9.2150,11.71750,24.080



DS14_GNDVI_Derived_Thesis.xlsx
Numerical columns:
['GNDVI', 'NDVI', 'EVI', 'LSWI', 'Canopy LAI est', 'Chlorophyll est ug cm2']


,count,mean,std,min,25%,50%,75%,max
GNDVI,120.0,0.544634,0.170521,0.1421,0.398500,0.5567,0.680900,0.8838
NDVI,120.0,0.607647,0.144737,0.2899,0.522275,0.6229,0.708650,0.8744
EVI,120.0,1.082322,0.912197,0.2702,0.567775,0.8565,1.180450,6.5818
LSWI,120.0,0.333221,0.179344,-0.1076,0.193925,0.3397,0.465225,0.6929
Canopy LAI est,120.0,3.265000,1.089185,0.6000,2.465000,3.3150,4.067500,5.9100
Chlorophyll est ug cm2,120.0,24.486250,7.835981,5.7000,19.000000,24.2550,30.417500,41.5200



DS17_Chlorophyll_Thesis.xlsx
Numerical columns:
['Block', 'Leaf age days', 'Chl a mg per gFW', 'Chl b mg per gFW', 'Total Chl mg per gFW', 'Chl a to b ratio', 'Carotenoid mg per gFW', 'Chl to Carotenoid ratio', 'SPAD value', 'Chlorophyll Index CI', 'Pheophytin mg per gFW', 'Chl degradation pct', 'Fv Fm PSII efficiency', 'Phi PSII quantum yield', 'NPQ non photochem quenching', 'Chl specific absorption', 'Light use efficiency LUE', 'GNDVI from DS14', 'N content pct from DS11', 'Nutrient balance NB DS16', 'Notes']


,count,mean,std,min,25%,50%,75%,max
Block,100.0,2.560000,1.103896,1.00000,2.000000,2.500000,4.000000,4.00000
Leaf age days,100.0,38.290000,16.049353,12.00000,24.000000,40.500000,51.000000,66.00000
Chl a mg per gFW,100.0,4.666724,0.359311,3.62210,4.483625,4.765450,5.000000,5.00000
Chl b mg per gFW,100.0,1.732999,0.232955,1.28680,1.583950,1.700450,1.897675,2.25860
Total Chl mg per gFW,100.0,6.399723,0.522867,4.96710,6.118650,6.537550,6.716825,7.25860
Chl a to b ratio,100.0,2.725310,0.301540,2.20200,2.434500,2.748500,3.017750,3.18700
Carotenoid mg per gFW,100.0,1.499813,0.219822,1.02720,1.340750,1.483600,1.649125,1.96790
Chl to Carotenoid ratio,100.0,4.330130,0.529780,3.57200,3.876250,4.251000,4.710250,5.52600
SPAD value,100.0,61.774700,7.811611,46.05000,55.920000,62.825000,68.510000,72.00000
Chlorophyll Index CI,100.0,0.772177,0.097647,0.57560,0.698975,0.785300,0.856350,0.90000


### A few important findings

Most datasets have no missing values in their meaningful measurement columns.
Notes in DS1 and DS17 is entirely empty, so we can ignore it.
DS17 has 100 observations, while the other datasets have 120. This is expected from our earlier ID check.
Some variables are essentially alternate representations of the same measurement. For example, N content pct DW and Kjeldahl N mg kg are directly related, and similarly for P and K. We should not blindly use every numerical column as a separate feature later.
DS14 EVI has a maximum of 6.5818, considerably above its 75th percentile of 1.18045. That is something we should investigate rather than automatically deleting.
DS2 TCH_t_per_ha and TSH_t_per_ha also have relatively high maximums compared with their medians, so we'll inspect their distributions before deciding whether they are genuine values or outliers.

### Check categorical variables

In [16]:
for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)
    categorical_cols = df.select_dtypes(include="object").columns
    for col in categorical_cols:
        print(f"\n{col}:")
        print("Unique values:", df[col].nunique())
        print(df[col].value_counts(dropna=False).head(20))


DS1_plot_registry.xlsx

Plot_ID:
Unique values: 120
Plot_ID
WC-001    1
WC-002    1
WC-003    1
WC-004    1
WC-005    1
WC-006    1
WC-007    1
WC-008    1
WC-009    1
WC-010    1
WC-011    1
WC-012    1
WC-013    1
WC-014    1
WC-015    1
WC-016    1
WC-017    1
WC-018    1
WC-019    1
WC-020    1
Name: count, dtype: int64

Village:
Unique values: 3
Village
Matiaria    41
Masjidwa    40
Sirsa       39
Name: count, dtype: int64

Season:
Unique values: 3
Season
2021-22    51
2022-23    37
2023-24    32
Name: count, dtype: int64

Irrigation_type:
Unique values: 3
Irrigation_type
Canal       64
Tubewell    39
Rainfed     17
Name: count, dtype: int64

Soil_type:
Unique values: 3
Soil_type
Alluvial      76
Sandy_loam    29
Clay_loam     15
Name: count, dtype: int64

Planting_date:
Unique values: 120
Planting_date
2021-11-01    1
2021-11-04    1
2021-11-07    1
2021-11-10    1
2021-11-13    1
2021-11-16    1
2021-11-19    1
2021-11-22    1
2021-11-25    1
2021-11-28    1
2021-12-01    1
202

### Check date columns

In [17]:
for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)
    date_candidates = [
        col for col in df.columns
        if "date" in col.lower()
    ]
    print("Date columns:", date_candidates)
    for col in date_candidates:
        print(f"\n{col}")
        print(df[col].head())


DS1_plot_registry.xlsx
Date columns: ['Planting_date', 'Harvest_date']

Planting_date
0    2021-11-01
1    2021-11-04
2    2021-11-07
3    2021-11-10
4    2021-11-13
Name: Planting_date, dtype: object

Harvest_date
0    2022-10-15
1    2022-10-18
2    2022-10-21
3    2022-10-24
4    2022-10-27
Name: Harvest_date, dtype: object

DS2_TSH_Records.xlsx
Date columns: []

DS6_Precipitaion.xlsx
Date columns: []

DS11_Nitrogen_Thesis.xlsx
Date columns: ['Sample date']

Sample date
0    2022-06-01
1    2022-06-04
2    2022-06-07
3    2022-06-10
4    2022-06-13
Name: Sample date, dtype: object

DS12_Phosphorus_Thesis.xlsx
Date columns: ['Sample date']

Sample date
0    2022-06-01
1    2022-06-04
2    2022-06-07
3    2022-06-10
4    2022-06-13
Name: Sample date, dtype: object

DS13_Potassium_Thesis.xlsx
Date columns: ['Sample date']

Sample date
0    2022-06-01
1    2022-06-04
2    2022-06-07
3    2022-06-10
4    2022-06-13
Name: Sample date, dtype: object

DS14_GNDVI_Derived_Thesis.xlsx
Date co

### Check duplicate Plot IDs

In [18]:
for name, df in datasets.items():

    duplicate_ids = df[df["Plot_ID"].duplicated(keep=False)]

    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)

    if duplicate_ids.empty:
        print("No duplicate Plot_IDs.")
    else:
        print("Duplicate Plot_IDs found:")
        display(duplicate_ids[["Plot_ID"]])


DS1_plot_registry.xlsx
No duplicate Plot_IDs.

DS2_TSH_Records.xlsx
No duplicate Plot_IDs.

DS6_Precipitaion.xlsx
No duplicate Plot_IDs.

DS11_Nitrogen_Thesis.xlsx
No duplicate Plot_IDs.

DS12_Phosphorus_Thesis.xlsx
No duplicate Plot_IDs.

DS13_Potassium_Thesis.xlsx
No duplicate Plot_IDs.

DS14_GNDVI_Derived_Thesis.xlsx
No duplicate Plot_IDs.

DS17_Chlorophyll_Thesis.xlsx
No duplicate Plot_IDs.


### Check potential outliers

In [19]:
for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(name)
    print("=" * 80)
    numeric_cols = df.select_dtypes(include="number").columns
    for col in numeric_cols:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower = Q1 - 1.5 * IQR
        upper = Q3 + 1.5 * IQR
        outliers = df[(df[col] < lower) | (df[col] > upper)]
        if len(outliers) > 0:
            print(
                f"{col}: {len(outliers)} potential outlier(s) "
                f"| Lower = {lower:.3f} | Upper = {upper:.3f}"
            )


DS1_plot_registry.xlsx

DS2_TSH_Records.xlsx
TCH_t_per_ha: 2 potential outlier(s) | Lower = -50.501 | Upper = 142.969
TSH_t_per_ha: 4 potential outlier(s) | Lower = -5.051 | Upper = 14.242

DS6_Precipitaion.xlsx
Precip_growth_phase_mm: 4 potential outlier(s) | Lower = 456.462 | Upper = 1478.963
Precip_Jun_Aug_mm: 4 potential outlier(s) | Lower = 308.813 | Upper = 879.912
Precip_Sep_Nov_mm: 4 potential outlier(s) | Lower = 122.263 | Upper = 424.962

DS11_Nitrogen_Thesis.xlsx

DS12_Phosphorus_Thesis.xlsx
P content pct DW: 2 potential outlier(s) | Lower = 0.087 | Upper = 0.412
P mg per kg DW: 2 potential outlier(s) | Lower = 867.000 | Upper = 4125.000
P2O5 pct: 2 potential outlier(s) | Lower = 0.199 | Upper = 0.945

DS13_Potassium_Thesis.xlsx
K Na ratio: 4 potential outlier(s) | Lower = 0.349 | Upper = 18.539

DS14_GNDVI_Derived_Thesis.xlsx
EVI: 12 potential outlier(s) | Lower = -0.351 | Upper = 2.099

DS17_Chlorophyll_Thesis.xlsx
Chl a mg per gFW: 2 potential outlier(s) | Lower = 3.709 

### Convert date columns

In [20]:
for name, df in datasets.items():
    date_cols = [
        col for col in df.columns
        if "date" in col.lower()
    ]
    for col in date_cols:
        datasets[name][col] = pd.to_datetime(
            datasets[name][col],
            errors="coerce"
        )
print("Date columns converted successfully.")

for name, df in datasets.items():

    date_cols = [
        col for col in df.columns
        if "date" in col.lower()
    ]

    if date_cols:
        print("\n" + name)
        print(df[date_cols].dtypes)

Date columns converted successfully.

DS1_plot_registry.xlsx
Planting_date    datetime64[ns]
Harvest_date     datetime64[ns]
dtype: object

DS11_Nitrogen_Thesis.xlsx
Sample date    datetime64[ns]
dtype: object

DS12_Phosphorus_Thesis.xlsx
Sample date    datetime64[ns]
dtype: object

DS13_Potassium_Thesis.xlsx
Sample date    datetime64[ns]
dtype: object

DS14_GNDVI_Derived_Thesis.xlsx
Acquisition date    datetime64[ns]
dtype: object

DS17_Chlorophyll_Thesis.xlsx
Sample date    datetime64[ns]
dtype: object


### Verify the analytical dataset size

In [21]:
plot_id_sets = {
    name: set(df["Plot_ID"])
    for name, df in datasets.items()
}

common_ids = set.intersection(*plot_id_sets.values())

print("Number of common Plot IDs:", len(common_ids))
print("First 10:", sorted(common_ids)[:10])
print("Last 10:", sorted(common_ids)[-10:])

Number of common Plot IDs: 100
First 10: ['WC-001', 'WC-002', 'WC-003', 'WC-004', 'WC-005', 'WC-006', 'WC-007', 'WC-008', 'WC-009', 'WC-010']
Last 10: ['WC-091', 'WC-092', 'WC-093', 'WC-094', 'WC-095', 'WC-096', 'WC-097', 'WC-098', 'WC-099', 'WC-100']


### The potential outliers are not automatically errors. We'll keep them for now.
Date columns were successfully converted to datetime64[ns].
Most importantly, we confirmed 100 common plots (WC-001–WC-100) across all 8 datasets.
The DS17 limitation is therefore clear: our integrated 8-dataset analysis will use these 100 plots.

### Merge DS1 + DS2 + DS6

In [22]:
master = datasets["DS1_plot_registry.xlsx"].copy()
print("Starting master shape:", master.shape)

Starting master shape: (120, 13)


In [23]:
master = master.merge(
    datasets["DS2_TSH_Records.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("_registry", "_tsh")
)

print("After DS2 merge:", master.shape)

After DS2 merge: (120, 26)


In [24]:
master = master.merge(
    datasets["DS6_Precipitaion.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_precip")
)

print("After DS6 merge:", master.shape)

After DS6 merge: (120, 41)


In [25]:
print("Number of columns:", len(master.columns))
print("\nColumns:")
for i, col in enumerate(master.columns, start=1):
    print(i, col)

Number of columns: 41

Columns:
1 Plot_ID
2 Village_registry
3 Block_registry
4 Season_registry
5 Latitude_deg
6 Longitude_deg
7 Plot_Area_ha
8 Irrigation_type
9 Soil_type
10 Planting_date
11 Harvest_date
12 Growth_days
13 Notes
14 Village_tsh
15 Season_tsh
16 Cultivar
17 Cycle
18 Block_tsh
19 TCH_t_per_ha
20 CCS_pct
21 TSH_t_per_ha
22 Sugar_recovery_pct
23 Juice_purity_pct
24 Brix_pct
25 Pol_pct
26 Mill_ID
27 Village
28 Season
29 Source
30 Grid_lat
31 Grid_lon
32 Precip_growth_phase_mm
33 Precip_Jun_Aug_mm
34 Precip_Sep_Nov_mm
35 Precip_Dec_Feb_mm
36 Dry_spell_days_gt5
37 Wet_days_gt10mm
38 Max_1day_precip_mm
39 Drought_flag
40 Excess_rain_flag
41 SPI_3month


In [26]:
print("\nRows:", len(master))
print("Unique Plot IDs:", master["Plot_ID"].nunique())

print("\nFirst 5 Plot IDs:")
print(master["Plot_ID"].head().tolist())

print("\nLast 5 Plot IDs:")
print(master["Plot_ID"].tail().tolist())


Rows: 120
Unique Plot IDs: 120

First 5 Plot IDs:
['WC-001', 'WC-002', 'WC-003', 'WC-004', 'WC-005']

Last 5 Plot IDs:
['WC-116', 'WC-117', 'WC-118', 'WC-119', 'WC-120']


### Merge DS11 — Nitrogen.

In [27]:
master = master.merge(
    datasets["DS11_Nitrogen_Thesis.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_N")
)

print("After DS11 merge:", master.shape)

After DS11 merge: (120, 54)


In [28]:
print("Number of columns:", len(master.columns))
for i, col in enumerate(master.columns, start=1):
    print(i, col)

Number of columns: 54
1 Plot_ID
2 Village_registry
3 Block_registry
4 Season_registry
5 Latitude_deg
6 Longitude_deg
7 Plot_Area_ha
8 Irrigation_type
9 Soil_type
10 Planting_date
11 Harvest_date
12 Growth_days
13 Notes
14 Village_tsh
15 Season_tsh
16 Cultivar
17 Cycle
18 Block_tsh
19 TCH_t_per_ha
20 CCS_pct
21 TSH_t_per_ha
22 Sugar_recovery_pct
23 Juice_purity_pct
24 Brix_pct
25 Pol_pct
26 Mill_ID
27 Village
28 Season
29 Source
30 Grid_lat
31 Grid_lon
32 Precip_growth_phase_mm
33 Precip_Jun_Aug_mm
34 Precip_Sep_Nov_mm
35 Precip_Dec_Feb_mm
36 Dry_spell_days_gt5
37 Wet_days_gt10mm
38 Max_1day_precip_mm
39 Drought_flag
40 Excess_rain_flag
41 SPI_3month
42 Village_N
43 Season_N
44 Sample date
45 Growth stage
46 Leaf position
47 N content pct DW
48 N class
49 Kjeldahl N mg kg
50 Total protein pct
51 Chlorophyll SPAD
52 N fertiliser kg ha
53 Method
54 Lab ID


### Merge DS12 phosphorus

In [29]:
master = master.merge(
    datasets["DS12_Phosphorus_Thesis.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_P")
)
print("After DS12 merge:", master.shape)
print("Rows:", len(master))
print("Unique Plot IDs:", master["Plot_ID"].nunique())

After DS12 merge: (120, 67)
Rows: 120
Unique Plot IDs: 120


### Merge DS13 potassium

In [30]:
master = master.merge(
    datasets["DS13_Potassium_Thesis.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_K")
)
print("After DS13 merge:", master.shape)
print("Rows:", len(master))
print("Unique Plot IDs:", master["Plot_ID"].nunique())

After DS13 merge: (120, 80)
Rows: 120
Unique Plot IDs: 120


### Merge DS14 (GNDVI / Remote Sensing)

In [31]:
master = master.merge(
    datasets["DS14_GNDVI_Derived_Thesis.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_GNDVI")
)
print("After DS14 merge:", master.shape)
print("Rows:", len(master))
print("Unique Plot IDs:", master["Plot_ID"].nunique())

After DS14 merge: (120, 93)
Rows: 120
Unique Plot IDs: 120


### Merge DS17
Now we'll add the detailed chlorophyll dataset.
Remember: DS17 only contains WC-001 to WC-100, so this merge should reduce the master from 120 rows to 100 rows.

In [32]:
master = master.merge(
    datasets["DS17_Chlorophyll_Thesis.xlsx"],
    on="Plot_ID",
    how="inner",
    suffixes=("", "_CHL")
)
print("After DS17 merge:", master.shape)
print("Rows:", len(master))
print("Unique Plot IDs:", master["Plot_ID"].nunique())

After DS17 merge: (100, 126)
Rows: 100
Unique Plot IDs: 100


### Final master dataset

In [33]:
print("=" * 60)
print("FINAL MASTER DATASET")
print("=" * 60)
print("Shape:", master.shape)
print("Rows:", len(master))
print("Columns:", len(master.columns))
print("Unique Plot IDs:", master["Plot_ID"].nunique())
print("\nPlot ID range:")
print(master["Plot_ID"].min(), "to", master["Plot_ID"].max())

FINAL MASTER DATASET
Shape: (100, 126)
Rows: 100
Columns: 126
Unique Plot IDs: 100

Plot ID range:
WC-001 to WC-100


### Check the final master for missing values

In [34]:
missing = master.isna().sum()
missing = missing[missing > 0].sort_values(ascending=False)
print("Columns containing missing values:")
if missing.empty:
    print("No missing values.")
else:
    print(missing)

Columns containing missing values:
Notes          100
Notes_CHL      100
Stress type     41
dtype: int64


### Save the master dataset

In [38]:
os.makedirs("../outputs/tables", exist_ok=True)
master_path = "../outputs/tables/Sugarcane_Master_Dataset.csv"
master.to_csv(master_path, index=False)
print("Master dataset saved successfully.")
print("Location:", master_path)
print("Shape:", master.shape)

Master dataset saved successfully.
Location: ../outputs/tables/Sugarcane_Master_Dataset.csv
Shape: (100, 126)
